Update language text to specify language versions. (#863)
dart-lang/language: accepted/future-releases/language-versioning/feature-specification.md

For breaking changes, the tool must be able to handle multiple different versions of the language, and do so in a way that allows libraries on different language versions  to interact in a predictable and useful way. For example, the "Null Safety" feature will allow libraries with non-nullable or safely nullable types *and* unmigrated libraries with unsafely nullable types, and those libraries can send values back and forth between them. This is where the main complication lies, and that is beyond the scope of this document which only specifies how we assign language versions to individual libraries.

This feature cannot support language changes that are fundamentally incompatible with earlier language versions, we do need the run-time system to be able to contain both language versions at the same time. For the same reason, this feature does not handle *platform library* changes that are not backwards compatible. The libraries are shared between the different language version libraries in a program, and must work for all of these. (Our tools should still detect use of library features not supported by the minimum required SDK version for a package.)

> Versioning uses _language version number_ only. You cannot opt-in to one feature of a new version, but not the rest. You either migrate to that new version, or you do not. This allows us to keep the combinatorial complexity down to being (roughly) linear in the number of supported versions.

Reviewer: ```suggestion
Versioning uses the language version number only. You cannot opt-in to one feature of a new version, but not the rest. You either migrate to that new version, or you do not. This allows us to keep the combinatorial complexity down to being (roughly) linear in the number of supported versions.
```

Again, removed italic property because this is not (no longer) the defining occurrence of that phrase.

### Experimental Languages

The actual language of the next Dart release is not known for certain until that Dart SDK is finally released. All new features currently being developed are experimental and subject to being changed, postponed, or withdrawn.

> During the development of a next Dart release, the SDK exists only as developer releases (for example, the Dart SDK version on bleeding edge is 2.8.0-dev.9.0). There is no Dart 2.8 language yet, but there is a family of experimental Dart 2.8 languages that users can experiment with.

Reviewer: ```suggestion
During the development of a next Dart release, the SDK exists only as developer releases (for example, the Dart SDK version on bleeding edge is 2.8.0-dev.9.0). There is no Dart 2.8 language version yet, but there is a family of experimental Dart 2.8 languages that users can experiment with.
```

It makes sense to introduce those 'experimental Dart 2.8 languages', and it would be confusing to insist that they are 'versions' of the language. Keeping them separate enables us to always have an associated version number whenever we have a language version.

> We switch between these language versions using the `–enable-experiment` flag to the Dart compilers and other tools. The current *base* Dart 2.8 language, the one used with no experiments enabled, is the same as the 2.7 language. A different 2.8 language can be used by enabling one or more experiments. There are, at time of writing, four experiments in various degrees of completion:

Reviewer: ```suggestion
We switch between these experimental languages using the `–enable-experiment` flag to the Dart compilers and other tools. The current *base* language for experiments under version 2.8, the one used with no experiments enabled, is the same as language version 2.7. A different 2.8 experimental language can be used by enabling one or more experiments. There are, at time of writing, four experiments in various degrees of completion:
```

It does so by:

- Assigning a default language version to each Dart package.
- Assigning each Dart library to at most one Dart package.
- Making each library inherit its package’s default language version unless it specifies a library-specific language version override.
- Defaulting to the most recent language version if there is no other hint.

The *package configuration* for a Dart program is stored in the new [`.dart_tool/package_config.json`](https://github.com/dart-lang/language/blob/master/accepted/future-releases/language-versioning/package-config-file-v2.md) file (which replaces `.packages` as the package configuration file). It contains information which allows Dart tools to resolve `package:` URIs to files, assign files to packages, and assign a *default* language version to each package. 

### Default Language Version

Each *package* in a program has a default language version associated with it, and each *library* in that package has that language version as its default language version. A library can override the default language version explicitly as described below.

> The package configuration file,  `.dart_tool/package_config.json`, file is a JSON formatted text file which defines the configuration for a number of named packages using entries of the form:

Reviewer: Whitespace:

```suggestion
The package configuration file, `.dart_tool/package_config.json`, file is a JSON formatted text file which defines the configuration for a number of named packages using entries of the form:
```

A file which does not belong to *any* package also uses the language version of the current SDK as its default language version. If there is no `package_config.json` file available (yet), all files are considered as not belonging to a package and default to the SDK language version.

The `package_config.json` file is, by default, generated by the Pub tool. It fills in the language version of each package using the language version of the minimum required SDK specified in that package’s `pubspec.yaml` file. This ensures that the default language version of libraries in a package are not later than the earliest language version that the package claims to be able to run on. The SDK requirement is optional, and if there is none, then the language version entry is also omitted (the `"languageVersion"` entry is optional). 

For a Pub package, the root directory will typically be the package directory, and the package URI root is the `lib/` directory inside the package directory. This means that all Dart libraries in, e.g., `test/` and `bin/` directories of a Pub package are considered as belonging to the package, not just the libraries accessed using `package:` URIs.

> Tools will not have a language default version available for any package until a `package_config.json` file has been generated by running `pub get`. Any tool able to understand Pub package layout *may* derive the information directly from the  `pubspec.yaml` of the current package, rather than rely on an entry in the `package_config.json` file, as long as it derives the same values that Pub would have written into the `package_config.json` file. A lack of a `package_config.json` file is only an issue for *new* packages, or packages with no dependencies, because otherwise the package is mostly unusable until you run `pub get` anyway. A new package is likely to want to be run at the most recent language version, so this is not a major problem. An IDE which reruns `pub get` after each change to `pubspec.yaml` will keep the `package_config.json` file up-to-date.

Reviewer: Whitespace:

```suggestion
Tools will not have a language default version available for any package until a `package_config.json` file has been generated by running `pub get`. Any tool able to understand Pub package layout *may* derive the information directly from the `pubspec.yaml` of the current package, rather than rely on an entry in the `package_config.json` file, as long as it derives the same values that Pub would have written into the `package_config.json` file. A lack of a `package_config.json` file is only an issue for *new* packages, or packages with no dependencies, because otherwise the package is mostly unusable until you run `pub get` anyway. A new package is likely to want to be run at the most recent language version, so this is not a major problem. An IDE which reruns `pub get` after each change to `pubspec.yaml` will keep the `package_config.json` file up-to-date.
```

- Allow language version markers above the package default level (for generated code).
- State that the version used from `pubspec.yaml` is the lower bound, whether inclusive or exclusive, and that the language version is the major/minor versions of that.

1.4:  Update to use new `package_config.json` file.

1.5: Rewrite and elaboration:

> - Make it explicit how the Pub tool choses the package default language version

Reviewer: ```suggestion
- Make it explicit how the Pub tool choses the package default language version.
```
